File: client/server/metadata_handler.py
import json
from flask import render_template


class MetadataHandler:
    def __init__(self):
        self.static_metadata = json.load(open("static/static_metadata.json"))

    def _get_index(self, metadata=None):
        if metadata is None:
            title = self.static_metadata["default"]["title"]
            description = self.static_metadata["default"]["description"]
            image = self.static_metadata["default"]["image"]
        else:
            title = metadata.get("title", self.static_metadata["default"]["title"])
            description = metadata.get(
                "description", self.static_metadata["default"]["description"]
            )
            image = metadata.get("image", self.static_metadata["default"]["image"])

        return render_template(
            "index.html", title=title, description=description, image=image
        )

    def render_index(self, path=None, metadata=None):
        if metadata is not None:
            return self._get_index(metadata)
        elif path is not None:
            for path_metadata in self.static_metadata["paths"]:
                if path_metadata["path"] == path:
                    return self._get_index(path_metadata)

        return self._get_index()

File: client/src/handlers/metadata_handler.ts
import staticMetadata from 'public/static_metadata.json'

interface DefaultMetadata {
    title: string;
    description: string;
    image: string;
}

interface Metadata extends Partial<DefaultMetadata> {}

interface PathMetadata extends Metadata {path: string}

export class MetadataHandler {
    static setMetadata(params: {
        path?: string,
        metadata?: Metadata
    }) {
        const metadata: DefaultMetadata = {...staticMetadata.default}
        if(params.metadata != null) {
            params.metadata.title && (metadata.title = params.metadata.title)
            params.metadata.description && (metadata.description = params.metadata.description)
            params.metadata.image && (metadata.image = params.metadata.image)
        } else if(params.path != null) {
            for(let metadataPath of staticMetadata.paths as PathMetadata[]) {
                if(metadataPath.path === params.path) {
                    metadataPath.title && (metadata.title = metadataPath.title)
                    metadataPath.description && (metadata.description = metadataPath.description)
                    metadataPath.image && (metadata.image = metadataPath.image)
                }
            }
        }

        document.title = metadata.title

        const metaElements = document.getElementsByTagName("meta")
        for(let element of metaElements) {
            const name = element.getAttribute("name") || element.getAttribute("property")
            if(name != null) {
                if(/title/.test(name)) {
                    element.setAttribute("content", metadata.title)
                } else if(/description/.test(name)) {
                    element.setAttribute("content", metadata.description)
                } else if(/image/.test(name)) {
                    element.setAttribute("content", metadata.image)
                }
            }
        }

    } 
}